added MATLAB docs (#76)
ubccr/ccrdocs: pages/howto/matlab.md

# MATLAB on the HPC Clusters

UB has a sitewide academic license for MATLAB along with all the toolboxes and CCR provides a few versions of MATLAB on our HPC Systems.
 
There are a few different methods for running MATLAB on CCR HPC Systems and the method you chose depends on the type of MATLAB calculatons you with to run. This guide will focus on the main methods in which users would run MATLAB at CCR.
These Methods include:

- Running MATLAB GUI though OpenOndemand
- Running MATLAB batch jobs on the Clusters through Slurm
- Running MATLAB interactively on the command line
- Using MATLAB Parallel Compute Server to submit jobs from within Matlab

This guide does not give information on using MATLAB itself, it is assumed that the audience is already familiar with using it. If you are looking for specific MATLAB related info their documentation is located [here](https://www.mathworks.com/help/matlab/).

> 

Reviewer: Recommend adding a section here on how to find MATLAB versions:

## Versions of MATLAB available at CCR  

Software is updated regularly at CCR so users need to know how to search the existing software repositories to find the available versions of software they're interested in running.  In the MATLAB OnDemand app, the form provides a drop down menu listing the available versions of MATLAB.  For those using the command line, you can search using the `module spider` command.  First, specify a [software release version](../software/releases) and then search for MATLAB:  

```
module load ccrsoft/2023.01
module spider matlab

matlab:
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
     Versions:
        matlab/2021b
        matlab/2023b
```

additionalSubmitArgs =

    '--ntasks=5 --cpus-per-task=1 --ntasks-per-core=1 --mem-per-cpu=8gb --qos=general-compute --clusters ub-hpc -p general-compute --mail-type=ALL --mail-user=testuser@buffalo.edu'
```

You can check the Slurm job state from within MATLAB:
```
>> job.State

ans =

    'queued'

>> 

```
After the job is completed the state will be Finished and you can check the output:

```
>> job.State

ans =

    'finished'

>> job.fetchOutputs{:}

ans =

    8.4850

>> 

```

The job ran in 8.485 seconds using four workers.  **Note** that these jobs will always request N+1 CPU cores, since one worker is required to manage the batch job and pool of workers. 
For example, a job that needs eight workers will request nine CPU cores.

### Configuring Slurm Integration on CCR MATLAB (GUI)

Configuring the Slurm integration using the MATLAB GUI is pretty straight forward. 

- Start up MATLAB through an Ondemand Session as described [here](#running-matlab-gui-though-openondemand)

- From the MATLAB Session Click on `Parallel` -> `Discover Clusters`

- Leave the check box in the `On your Network` and click `Next`

> - Highlight the default `ub-hpc` clister and click `Finish`

Reviewer: ```suggestion
- Highlight the default `ub-hpc` cluster and click `Finish`
```

![](../images/matlab2.png)

You now have a default Cluster Profile that should work out of the box, but you will want to change some of the default settings such as user email settings.

To make changes to the default Cluster Profile:

- From the MATLAB Session Click on `Parallel` -> `Create and Manage Clusters`

- Click on the ub-hpc Cluster Profile and Click `Edit`

- Make your changes and click `Done`

### Configuring Remote Slurm Integration (GUI)

> It is possible to configure your local MATLAB to submit jobs to CCR from your Desktop or Workstation. The setup is very similar to Configuring Slurm Integration on CCR MATLAB once you have the correct integration scripts.

Reviewer: ```suggestion
It is possible to configure your local MATLAB to submit jobs to CCR from your desktop or workstation. The setup is very similar to [Configuring Slurm Integration on CCR MATLAB](#configuring-slurm-integration-on-ccr-matlab-command-line) once you have the correct integration scripts.
```

> NOTE: This proceedure assumes that you have MATLAB R2023b (Currently the only supported Version at CCR) installed on your local machine.

Reviewer: ```suggestion
NOTE: This procedure assumes that you have MATLAB R2023b (Currently the only supported version at CCR) installed on your local machine.
```

!!! Warning "SSH Key Credentials"
    Submission to the cluster requires SSH credentials.  You will be prompted for username and password or identity file (private key).  The username and location of the private key will be stored in MATLAB for future sessions.

> You will need do download the MATLAB CCR Slurm integration script and place them on your local machine in the MATLAB

Reviewer: ```suggestion
You will need to download the MATLAB CCR Slurm integration script and place them on your local machine in the MATLAB
```
...seems like this sentence isn't complete?

Download the [CCR Matlab Integration Scripts](https://g-ac407a.8c185.08cc.data.globus.org/R2023b/University-at-Buffalo.zip) and start MATLAB.  The ZIP file should be unzipped in the location returned by calling…

```
>> userpath

ans =

    '/home/testuser/Documents/MATLAB'

>> 

```

> Configure MATLAB to run parallel jobs on the cluster by using the configCluster This only needs to be run  once per version of MATLAB.

Reviewer: ```suggestion
Configure your local MATLAB to run parallel jobs on the cluster by using the `configCluster` cluster command.  This only needs to be run once per version of MATLAB.
```

```
>> configCluster
Username on UB-HPC (e.g. jdoe): >>testuser
Complete.  Default cluster profile set to "ub-hpc R2023b".
>> 

```

> You will need to configure your user infomation into the Profile including the location to your public SSH key:

Reviewer: Should this be *private* SSH key?

Author: yes. this is your private key. I will change

```
>> c = parcluster;
>> c.AdditionalProperties.EmailAddress = 'testuser@buffalo.edu'
>> c.AdditionalProperties.IdentityFile = '/home/testuser/.ssh/private.key'
```

> **NOTE:** We recommend setting the RemoteJobStorageLocation to your project space rather than your home directory.

Reviewer: ```suggestion
**NOTE:** We recommend setting the RemoteJobStorageLocation to your group's project directory, if you have one, rather than your home directory to prevent going over quota in your home directory.  
```

     Versions:
        matlab/2021b
        matlab/2023b
```

## Running MATLAB GUI though OpenOndemand 

The easiest way to use MATLAB on the HPC clusters is through the [CCR OnDemand Portal](../portals/ood.md). This method provides a more interactive use of MATLAB but requires a hands on approach. If you are looking for something more automated and scripted [see below](#running-matlab-batch-jobs-on-the-clusters-through-slurm)

To begin a session, click on "Interactive Apps" and then "MATLAB GUI". By default, MATLAB sessions are allocated 1 CPU, 2.8GB of RAM and 24 hours wall time (run time). You can override the defaults, however unless you are sure that your script has been explicitly parallelized using, for example, the Parallel Computing Toolbox (see below) leave the "Number of cores" set to 1 . Click "Launch" and then when your session is ready click "Launch MATLAB GUI". Note that the more resources you request, the more you will have to wait for your session to become available.

> ![](../images/matlab1.png)

Reviewer: ```suggestion
```
